cpp/box/Box.h

```diff
@@ -381,6 +382,53 @@ class Box
         });
     }
 
+    //! Calculate distance between two points using boundary conditions
+    /*! \param p_i Particle position
+        \param p_j Particle position to calculate distances from
+    */
+    float computeDistance(const vec3<float>& p_i, const vec3<float>& p_j) const
+    {
+            vec3<float> v_ij = wrap(p_j - p_i);
+            return std::sqrt(dot(v_ij, v_ij));
+    }
+
+    //! Calculate distance between a set of points and query points using boundary conditions
+    /*! \param points Particle positions
+        \param query_points Particle position to calculate distances from
+        \param dist Distances between points and query_points
+        \param Nvecs The number of points to calculate distances between
+    */
+    void computeDistances(vec3<float>* points, vec3<float>* query_points, float *dist, unsigned int Nvecs) const
+    {
+        util::forLoopWrapper(0, Nvecs, [=](size_t begin, size_t end) {
+            for (size_t i = begin; i < end; ++i)
+            {
+                dist[i] = computeDistance(points[i], query_points[i]);
+            }
+        });
+    }
+
+    //! Calculate distance between a set of points and query points using boundary conditions
+    /*! \param points Particle positions
+        \param query_points Particle position to calculate distances from
+    */
+    void computeAllDistances(vec3<float>* points, vec3<float>* query_points,
+        float* dist, unsigned int Nvecs, unsigned int Mvecs) const
+    {
+        util::forLoopWrapper(0, Nvecs, [=](size_t begin_n, size_t end_n) {
+            for (size_t i = begin_n; i < end_n; ++i)
+            {
+                util::forLoopWrapper(0, Mvecs, [=](size_t begin_m, size_t end_m) {
```

Comment:
The `util::forLoopWrapper` is for parallelization (though the name isn't clear about that). I believe this nesting of the wrapper will trigger parallel execution inside of parallel execution, which sounds like a bad idea. I think the inner loop should be a plain `for`.

Reply:
Actually nesting parallel loops like this is perfectly safe to do, especially since each iteration accesses a different index. In practice, we may not see much in the way of performance gains, but there's no harm in doing so. I could imagine use-cases where `Mvecs >> NVecs`, in which case this could be beneficial.

Reply:
Okay, it seems like this is supported and would probably be good enough as it is. Ideally we would replace this with something like `tbb::blocked_range2d` for the best performance. See references:

https://stackoverflow.com/questions/29719787/parallel-more-than-one-nested-loops-with-tbb
https://software.intel.com/en-us/node/506147

Reply:
Yes, I thought about this as well, and I agree that `blocked_range2d` is potentially a more performant option. However, since it makes the code less readable I'd prefer not to prematurely optimize until we encounter meaningful performance hits with this version of the code.

Reply:
I don't think it's any less readable and I don't think it's premature to change since it's a pretty simple alteration and it's an O(N^2) algorithm... but I won't hold up this PR for that.

Reply:
@klywang up to you if you would like to give this a shot. It shouldn't be too difficult to switch to using a `blocked_range2d`, but let me or Bradley know if you have trouble with it. I find it a bit less readable than for loops, but it should be faster in general so worth a shot if you're up for it.

Reply:
I made a new function(?) in `utils.h` called `ForLoopWrapper2D` using `blocked_range2d`. I can move it out of `utils.h` and implement it directly in `Box::computeAllDistances` if needed.

Reply:
I think this was a good move, let's move forward with that function for now.